autorest/azure/sender.go

```diff
@@ -0,0 +1,187 @@
+package azure
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/Azure/go-autorest/autorest"
+)
+
+// DoRetryForStatusCodes is the Azure specific implementation for retring requests
+func DoRetryForStatusCodes(azc Client, codes ...int) autorest.SendDecorator {
+	return func(s autorest.Sender) autorest.Sender {
+		return autorest.SenderFunc(func(r *http.Request) (resp *http.Response, err error) {
+			rr := autorest.NewRetriableRequest(r)
+			for attempt := 0; attempt <= azc.RetryAttempts; attempt++ {
+				err = rr.Prepare()
+				if err != nil {
+					return resp, err
+				}
+
+				resp, err = s.Do(rr.Request())
+				if err != nil {
+					return resp, err
+				}
+
+				if resp.StatusCode == http.StatusConflict {
+					var re RequestError
+					err = autorest.Respond(
+						resp,
+						autorest.ByUnmarshallingJSON(&re),
+						autorest.ByClosing(),
+					)
+					if err != nil {
+						return resp, err
+					}
+
+					if re.ServiceError != nil && re.ServiceError.Code == "MissingSubscriptionRegistration" {
+						err = register(azc, s, r, re)
+						if err != nil {
+							return resp, fmt.Errorf("failed auto registering Resource Provider: %s", err)
+						}
+					}
+				}
+				if !autorest.ResponseHasStatusCode(resp, codes...) {
+					return resp, err
+				}
+				delayed := autorest.DelayWithRetryAfter(resp, r.Cancel)
+				if !delayed {
+					autorest.DelayForBackoff(azc.RetryDuration, attempt, r.Cancel)
+				}
+			}
+			return resp, err
+		})
+	}
+}
+
+func getProvider(re RequestError) (string, error) {
+	if re.ServiceError != nil {
+		if re.ServiceError.Details != nil && len(*re.ServiceError.Details) > 0 {
+			detail := (*re.ServiceError.Details)[0].(map[string]interface{})
+			return detail["target"].(string), nil
+		}
+	}
+	return "", errors.New("provider was not found in the response")
+}
+
+func register(azc Client, s autorest.Sender, originalReq *http.Request, re RequestError) error {
```

Comment:
I would turn this into a method where (azc Client) is the receiver.  Also doesn't the client implement autorest.Sender?  If so then it seems you could remove this param.

Reply:
I need the client to get authorization again

Reply:
Sorry I meant remove the autorest.Sender param.

Reply:
I use that one for the test :(